crates/sui-json-rpc/src/authority_state.rs

```diff
@@ -0,0 +1,654 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::anyhow;
+use arc_swap::Guard;
+use async_trait::async_trait;
+use jsonrpsee::core::Error as RpcError;
+use jsonrpsee::types::{error::CallError, ErrorObject};
+use move_core_types::language_storage::TypeTag;
+use mysten_metrics::spawn_monitored_task;
+use std::collections::{BTreeMap, HashMap};
+use std::sync::Arc;
+use sui_core::authority::authority_per_epoch_store::AuthorityPerEpochStore;
+use sui_core::authority::{AuthorityState, AuthorityStore};
+use sui_core::subscription_handler::SubscriptionHandler;
+use sui_json_rpc_types::{
+    Checkpoint, Coin as SuiCoin, DevInspectResults, DryRunTransactionBlockResponse, EventFilter,
+    SuiEvent, SuiObjectDataFilter, TransactionFilter,
+};
+use sui_storage::indexes::TotalBalance;
+use sui_types::base_types::{
+    MoveObjectType, ObjectID, ObjectInfo, ObjectRef, SequenceNumber, SuiAddress,
+};
+use sui_types::committee::{Committee, EpochId};
+use sui_types::digests::{ChainIdentifier, TransactionDigest, TransactionEventsDigest};
+use sui_types::dynamic_field::DynamicFieldInfo;
+use sui_types::effects::{TransactionEffects, TransactionEvents};
+use sui_types::error::{SuiError, UserInputError};
+use sui_types::event::EventID;
+use sui_types::governance::StakedSui;
+use sui_types::messages_checkpoint::{
+    CheckpointContents, CheckpointContentsDigest, CheckpointDigest, CheckpointSequenceNumber,
+    VerifiedCheckpoint,
+};
+use sui_types::object::{Object, ObjectRead, PastObjectRead};
+use sui_types::storage::WriteKind;
+use sui_types::sui_serde::BigInt;
+use sui_types::sui_system_state::SuiSystemState;
+use sui_types::transaction::{TransactionData, TransactionKind, VerifiedTransaction};
+use thiserror::Error;
+use tokio::task::JoinError;
+
+#[cfg(test)]
+use mockall::automock;
+
+use crate::ObjectProvider;
+
+pub type StateReadResult<T = ()> = Result<T, StateReadError>;
+
+/// State trait to capture subset of AuthorityState used by CoinReadApi
+/// This allows us to also mock AuthorityState for testing
+#[cfg_attr(test, automock)]
+#[async_trait]
+pub trait State: Send + Sync {
```

Comment:
It seems like all the associated functions here have to do with requesting data, maybe a name like `StateRead` would be fitting